MetadataExtractor/IO/RandomAccessStream.cs

```diff
@@ -0,0 +1,578 @@
+﻿// Copyright (c) Drew Noakes and contributors. All Rights Reserved. Licensed under the Apache License, Version 2.0. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using JetBrains.Annotations;
+
+namespace MetadataExtractor.IO
+{
+    /// <summary>Reads and buffers data in chunks and provides methods for reading data types</summary>
+    /// <remarks>
+    /// This class implements buffered reading of data typically for use with <see cref="ReaderInfo"/> 
+    /// objects and provides methods for reading data types from it. Data is captured in configurable 
+    /// chunks for efficiency. Both seekable and non-seekable streams are supported.
+    /// </remarks>
+    /// <author>Kevin Mott https://github.com/kwhopper</author>
+    /// <author>Drew Noakes https://drewnoakes.com</author>
+    public class RandomAccessStream
+    {
+        public const long UnknownLengthValue = long.MaxValue;
+
+        private Stream? p_inputStream;
+        private long p_streamLength = -1;
+
+        //private readonly List<ReaderInfo> rdrList = new List<ReaderInfo>();
+        private bool p_isStreamFinished;
+
+        private const int DefaultChunkLength = 4 * 1024;
+        private readonly int p_chunkLength;
+        public Dictionary<long, byte[]> p_chunks = new Dictionary<long, byte[]>();
```

Comment:
Why `Dictionary<,>` and not `List<>`?

Reply:
Because chunks aren't necessarily read and stored in a linear sequence. There could be gaps so a Dictionary keyed on the chunk starting position is ideal. If the stream is seekable, note that BytesAvailable will first seek the stream to the desired 'chunkable' position and only read that chunk into the Dictionary; the chunk index becomes the key. This optimization ignores caching all the bytes in between, and is the main reason this implementation is as good or better than the current one on a seekable stream.

For example, suppose you have a FileStream (which is seekable) and the first chunk of the file has everything you need at the moment to start parsing. The Dictionary will have a key of "0" with a byte array the size of the 0-th chunk of the file. Maybe it tells you that the next IFD is 300,000 bytes further into the file. BytesAvailable will Seek to the closest chunk that contains this new position and read that one. Now there will be only two entries in the Dictionary - "0" and "73" if the chunk length is 4096 (assuming my math is right).

So, this avoids caching ~300,000 bytes of unneeded data in the Dictionary. The only way to know whether this chunk is already there is to key it by chunk index relative to the entire stream -- and we're then back to why a keyed structure is better. That isn't to say we might eventually read some or all of those bytes, but at this particular point they aren't needed.

This structure also works just fine if you're reading the whole thing linearly, such as with a NetworkStream.

Reply:
Thanks. Using your example, indexes 1 through 72 in a `List<byte[]>` could be `null` which would take up a trivial amount of space. This would then mean that lookups in the list would be faster (simple pointer arithmetic) than a dictionary lookup.

This kind of thing is why I think it's time to get some benchmarks set up. Performance is a very important feature of this library, particularly in the IO area.

BenchmarkDotNet is an amazing tool and very easy to use. If you haven't tried it before, I think you'll really enjoy it. This page has some copy/paste-able code to start from:

https://benchmarkdotnet.org/articles/guides/getting-started.html

I'd also add `[MemoryDiagnoser]` to the benchmark class, which will cause memory usage/GC stats to be included in the output.

It's probably best to create a new project in the solution for these benchmarks. I am happy to help with this.